File: ios/CareLog/CareLog/Upload/UploadViewModel.swift
import Foundation
import SwiftUI
import PhotosUI

/// ViewModel for upload screen.
@MainActor
final class UploadViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var selectedFileType: FileType = .medicalPhoto
    @Published var isUploading: Bool = false
    @Published var uploadProgress: Double?
    @Published var uploadSuccess: Bool = false
    @Published var error: String?

    // MARK: - Dependencies

    private let uploadService = UploadService.shared

    // MARK: - Upload Methods

    func uploadDocument(url: URL) async {
        isUploading = true
        error = nil

        do {
            // Start accessing security-scoped resource
            guard url.startAccessingSecurityScopedResource() else {
                throw UploadError.invalidData
            }
            defer { url.stopAccessingSecurityScopedResource() }

            let data = try Data(contentsOf: url)
            let contentType = url.pathExtension == "pdf" ? "application/pdf" : "image/jpeg"

            _ = try await uploadService.uploadFile(
                data: data,
                fileType: .prescription,
                contentType: contentType,
                description: "Prescription"
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func uploadPrescriptionData(data: Data, contentType: String) async {
        isUploading = true
        error = nil

        do {
            _ = try await uploadService.uploadFile(
                data: data,
                fileType: .prescription,
                contentType: contentType,
                description: "Prescription"
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func uploadPhotoPickerItem(_ item: PhotosPickerItem) async {
        isUploading = true
        error = nil

        do {
            guard let data = try await item.loadTransferable(type: Data.self) else {
                throw UploadError.invalidData
            }

            _ = try await uploadService.uploadFile(
                data: data,
                fileType: selectedFileType,
                contentType: "image/jpeg",
                description: selectedFileType.displayName
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func uploadImage(_ image: UIImage) async {
        isUploading = true
        error = nil

        do {
            _ = try await uploadService.uploadImage(
                image: image,
                fileType: selectedFileType,
                description: selectedFileType.displayName
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func uploadAudio(url: URL) async {
        isUploading = true
        error = nil

        do {
            let data = try Data(contentsOf: url)

            _ = try await uploadService.uploadFile(
                data: data,
                fileType: .voiceNote,
                contentType: "audio/mp4",
                description: "Voice Note"
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func uploadVideo(url: URL) async {
        isUploading = true
        error = nil

        do {
            let data = try Data(contentsOf: url)

            _ = try await uploadService.uploadFile(
                data: data,
                fileType: .videoNote,
                contentType: "video/mp4",
                description: "Video Note"
            ) { [weak self] progress in
                Task { @MainActor in
                    self?.uploadProgress = progress
                }
            }

            isUploading = false
            uploadProgress = nil
            uploadSuccess = true

        } catch {
            isUploading = false
            uploadProgress = nil
            self.error = error.localizedDescription
        }
    }

    func clearError() {
        error = nil
    }
}

File: android/app/src/main/java/com/carelog/ui/upload/UploadViewModel.kt
package com.carelog.ui.upload

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.upload.FileType
import com.carelog.upload.UploadResult
import com.carelog.upload.UploadService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for upload screen.
 */
@HiltViewModel
class UploadViewModel @Inject constructor(
    private val uploadService: UploadService
) : ViewModel() {

    private val _uiState = MutableStateFlow(UploadUiState())
    val uiState: StateFlow<UploadUiState> = _uiState.asStateFlow()

    fun setSelectedFileType(type: FileType) {
        _uiState.update { it.copy(selectedFileType = type) }
    }

    fun uploadDocument(context: Context, uri: Uri) {
        viewModelScope.launch {
            _uiState.update { it.copy(isUploading = true, error = null) }

            val contentType = context.contentResolver.getType(uri) ?: "application/pdf"

            val result = uploadService.uploadFile(
                uri = uri,
                fileType = FileType.PRESCRIPTION,
                contentType = contentType,
                description = "Prescription"
            ) { progress ->
                _uiState.update { it.copy(uploadProgress = progress) }
            }

            when (result) {
                is UploadResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            uploadSuccess = true,
                            uploadProgress = null
                        )
                    }
                }
                is UploadResult.Error -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            error = result.message,
                            uploadProgress = null
                        )
                    }
                }
            }
        }
    }

    fun uploadImage(context: Context, uri: Uri, fileType: FileType) {
        viewModelScope.launch {
            _uiState.update { it.copy(isUploading = true, error = null) }

            val contentType = context.contentResolver.getType(uri) ?: "image/jpeg"

            val result = uploadService.uploadFile(
                uri = uri,
                fileType = fileType,
                contentType = contentType,
                description = fileType.displayName
            ) { progress ->
                _uiState.update { it.copy(uploadProgress = progress) }
            }

            when (result) {
                is UploadResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            uploadSuccess = true,
                            uploadProgress = null
                        )
                    }
                }
                is UploadResult.Error -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            error = result.message,
                            uploadProgress = null
                        )
                    }
                }
            }
        }
    }

    fun uploadAudio(uri: Uri) {
        viewModelScope.launch {
            _uiState.update { it.copy(isUploading = true, error = null) }

            val result = uploadService.uploadFile(
                uri = uri,
                fileType = FileType.VOICE_NOTE,
                contentType = "audio/mp4",
                description = "Voice Note"
            ) { progress ->
                _uiState.update { it.copy(uploadProgress = progress) }
            }

            when (result) {
                is UploadResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            uploadSuccess = true,
                            uploadProgress = null
                        )
                    }
                }
                is UploadResult.Error -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            error = result.message,
                            uploadProgress = null
                        )
                    }
                }
            }
        }
    }

    fun uploadVideo(uri: Uri) {
        viewModelScope.launch {
            _uiState.update { it.copy(isUploading = true, error = null) }

            val result = uploadService.uploadFile(
                uri = uri,
                fileType = FileType.VIDEO_NOTE,
                contentType = "video/mp4",
                description = "Video Note"
            ) { progress ->
                _uiState.update { it.copy(uploadProgress = progress) }
            }

            when (result) {
                is UploadResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            uploadSuccess = true,
                            uploadProgress = null
                        )
                    }
                }
                is UploadResult.Error -> {
                    _uiState.update {
                        it.copy(
                            isUploading = false,
                            error = result.message,
                            uploadProgress = null
                        )
                    }
                }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    fun resetUploadSuccess() {
        _uiState.update { it.copy(uploadSuccess = false) }
    }
}

/**
 * UI state for upload screen.
 */
data class UploadUiState(
    val selectedFileType: FileType = FileType.MEDICAL_PHOTO,
    val isUploading: Boolean = false,
    val uploadProgress: Float? = null,
    val uploadSuccess: Boolean = false,
    val error: String? = null
)
